# Sistema de Recomendação de Livros

Importando bibliotecas

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

ModuleNotFoundError: No module named 'pandas'

Leitura do arquivo de livros. Sep é o separador do arquivo, on_bad_lines ignora as linhas que não estão no formato correto, encoding é o tipo de codificação do arquivo e low_memory é para que o pandas não precise alocar muita memória para ler o arquivo.

In [47]:
books = pd.read_csv('BX-Books.csv', sep=';',on_bad_lines='skip',encoding="latin-1", low_memory=False)
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [48]:
books.shape

(271360, 8)

É um dataset com 271.360 livros e 8 atributos.

In [49]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

- ISBN: Número de identificação do livro
- Book-Title: Título do livro
- Book-Author: Autor do livro
- Year-Of-Publication: Ano de publicação do livro
- Publisher: Editora do livro
- Image-URL-S: URL da imagem da capa do livro (pequena)
- Image-URL-M: URL da imagem da capa do livro (média)
- Image-URL-L: URL da imagem da capa do livro (grande)

Vamos coletar apenas algumas colunas do dataset.

In [50]:
books = books[['ISBN',
                'Book-Title',
                  'Book-Author',
                    'Year-Of-Publication',
                      'Publisher',
                        'Image-URL-L']]

Renomeando algumas colunas

In [51]:
books.rename(columns = {'Book-Title':'title',
                         'Book-Author':'author',
                           'Year-Of-Publication':'year',
                             'Publisher':'publisher',
                             'Image-URL-L':'image_url'
                             }, inplace=True)

Agora temos nosso dataset contendo informações dos livros

In [52]:
books.head()

ISBN                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  \
0    Mark P. O. Morford  2002     Oxford University Press   
1  Richard Bruce Wright  2001       HarperFlamingo Canada   
2          Carlo D'Este  1991             HarperPerennial   
3      Gina Bari Kolata  1999        Farrar Straus Giroux   
4       E. J. W. Barber  1999  W. W. Norton &amp; Company   

                                           image_url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

Dataset usuários é um conjunto de dados com 278.858 usuários e 3 atributos.

In [53]:
users = pd.read_csv('BX-Users.csv', sep=';',on_bad_lines='skip',encoding="latin-1", low_memory=False)
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [54]:
users.shape

(278858, 3)

In [55]:
users.columns

Index(['User-ID', 'Location', 'Age'], dtype='object')

- User-ID: Número de identificação do usuário
- Location: Localização do usuário
- Age: Idade do usuário

Dataset avaliações é um conjunto de dados com 1.149.780 avaliações e 3 atributos.

In [56]:
ratings = pd.read_csv('BX-Book-Ratings.csv', sep=';',on_bad_lines='skip',encoding="latin-1", low_memory=False)
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [57]:
ratings.shape

(1149780, 3)

1149780 registros e 3 colunas

In [58]:
ratings.columns

Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')

- User-ID: Número de identificação do usuário
- ISBN: Número de identificação do livro
- Book-Rating: Avaliação do livro

In [59]:
print(books.shape)
print(users.shape)
print(ratings.shape)

(271360, 6)
(278858, 3)
(1149780, 3)


In [60]:
len(books["ISBN"].unique())

271360

O fato de avaliações ter muitos registros é porque cada usuário pode ter avaliado vários livros.

In [61]:
ratings.rename(columns={'User-ID':'user_id','Book-Rating':'rating'},inplace=True)

Vamos ver quantas vezes os usuários avaliaram os livros

In [62]:
ratings['user_id'].value_counts()

user_id
11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: count, Length: 105283, dtype: int64

Percebemos que há muitos usuários que avaliaram apenas um livro e um total de 105283 avaliações de livros diferentes.

Vamos levar em consideração apenas usuários que votaram mais de 200 vezes, pois assim os usuários que votaram apenas uma vez não terão muito peso na recomendação.

In [63]:
x = ratings['user_id'].value_counts()>200
x[x].shape

(899,)

899 usuários fizeram mais de 200 avaliações

In [64]:
y = x[x].index
ratings = ratings[ratings['user_id'].isin(y)]
ratings.shape

(526356, 3)

In [65]:
print('Média de avaliações por usuário: ',ratings.groupby('user_id')['rating'].count().mean().round())

Média de avaliações por usuário:  585.0


Agora vamos levar esses 899 usuários para o dataset de avaliações. No final, o dataset de recomendação terá seu número de registros reduzido para 526356, logicamente devido a redução de usuários. Cada usuário avaliou em média 585 livros.

Agora que temos os registros de avaliações dos usuários que avaliaram mais de 200 vezes, vamos incluir esses dados no dataframe livros, onde obteremos as informações dos livros que esses usuários avaliaram. 

In [87]:
ratings['ISBN'].nunique()

207291

In [93]:
ratings.head()

user_id        ISBN  rating
1456   277427  002542730X      10
1457   277427  0026217457       0
1458   277427  003008685X       8
1459   277427  0030615321       0
1460   277427  0060002050       0

In [95]:
books.shape

(271360, 6)

In [97]:
ratings.shape

(526356, 3)

In [66]:
ratings_with_books = ratings.merge(books, on='ISBN')
ratings_with_books.head()

user_id        ISBN  rating  \
0   277427  002542730X      10   
1     3363  002542730X       0   
2    11676  002542730X       6   
3    12538  002542730X      10   
4    13552  002542730X       0   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   publisher  \
0  John Wiley &amp; Sons Inc   
1  John Wiley &amp; Sons Inc   
2  John Wiley &amp; Sons Inc   
3  John Wiley &amp; Sons Inc   
4  John Wiley &amp; Sons Inc   

                                           image_url  
0  http://images.amazon.com/images/P/002542730X.0...  
1  http://images.amazon.com/images/P/002542730X.0...  
2  http://images.amazon.com/images/P/002542730X.0...  
3  http://images.amazon.com/images/P/002542730X.0...  
4  http://images.amazon.com/images/P/002542730X.0...

Agora uma nova coluna foi adicionada ao dataframe, isso aumentou o seu tamanho, pois agora um livro terá mais de uma avaliação por usuário, sendo cada registro

In [86]:
ratings_with_books.shape

(487671, 8)

In [68]:
ratings_with_books["title"].nunique()

160269

Temos agora um total de 160269 livros, pois são os livros que os usuários que avaliaram mais de 200 vezes avaliaram.

O número de registros original do dataset livros era de 271360, agora com as avaliações dos usuários que avaliaram mais de 200 vezes, o número de livros diminiu para 160269, mas o toal de registros aumentou para 487671, pois cada livro terá mais de uma avaliação por usuário.

Agora vamos obter a informação de quantas avaliações cada livro recebeu.

In [69]:
num_rating = ratings_with_books.groupby('title')['rating'].count().sort_values(ascending=False).reset_index()
num_rating.rename(columns={'rating':'num_of_ratings'},inplace=True)
num_rating.head()

title  num_of_ratings
0                Wild Animus             363
1      Bridget Jones's Diary             277
2  The Lovely Bones: A Novel             270
3               The Notebook             241
4          The Pelican Brief             236

Agora vamos incluir uma coluna de contagem de avaliações ao nosso dataframe

In [70]:
final_rating = ratings_with_books.merge(num_rating, on='title')
final_rating.head(2)

user_id        ISBN  rating  \
0   277427  002542730X      10   
1     3363  002542730X       0   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   publisher  \
0  John Wiley &amp; Sons Inc   
1  John Wiley &amp; Sons Inc   

                                           image_url  num_of_ratings  
0  http://images.amazon.com/images/P/002542730X.0...              82  
1  http://images.amazon.com/images/P/002542730X.0...              82

Vamos ver apenas livros com mais de 50 avaliações

In [71]:
final_rating = final_rating[final_rating['num_of_ratings']>=50]

In [72]:
final_rating.shape

(61853, 9)

Vamos eliminar usuários que votaram mais de uma vez no mesmo livro

In [73]:
final_rating.drop_duplicates(['user_id','title'],inplace=True)
final_rating.shape

(59850, 9)

2000 registros duplicados removidos

In [74]:
final_rating["title"].nunique(), final_rating["user_id"].nunique()

(742, 888)

In [75]:
final_rating.head()

user_id        ISBN  rating  \
0   277427  002542730X      10   
1     3363  002542730X       0   
2    11676  002542730X       6   
3    12538  002542730X      10   
4    13552  002542730X       0   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   publisher  \
0  John Wiley &amp; Sons Inc   
1  John Wiley &amp; Sons Inc   
2  John Wiley &amp; Sons Inc   
3  John Wiley &amp; Sons Inc   
4  John Wiley &amp; Sons Inc   

                                           image_url  num_of_ratings  
0  http://images.amazon.com/images/P/002542730X.0...              82  
1  http://images.amazon.com/images/P/002542730X.0...              82  
2  http://images.amazon.com/images/P/002542730X.0...              82  
3  http://images.amazon.com/images/P/002542730X.0...              82  
4  http://images.amazon.com/images/P/002542730X.0...              82

Esse é o nosso dataframe final, composto por 742 livros e 888 usuários, ou seja, 59850 registros de livros com mais de 50 avaliações, avaliados por usuários que avaliaram mais de 200 vezes.

Tabela pivot

O que faremos agora é um pivot table, onde os índices serão os usuários, as colunas serão os livros e os valores serão as avaliações. Isso porque estamos interessados em saber quais livros cada usuário avaliou, e não qual livro foi avaliado por cada usuário.

Para isso, cada coluna será baseada nos valores de user_id, os índices serão os títulos de cada livro e os valores serão as avaliações que cada usuário deu para cada livro.

In [76]:
book_pivot = final_rating.pivot_table(columns='user_id',index='title',values='rating')
book_pivot.fillna(0,inplace=True)
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [77]:
book_pivot.shape

(742, 888)

Vamos agora fazer o uso da biblioteca scipy para transformar nossa matriz pivot em uma matriz esparsa. Isso porque a matriz pivot é muito grande e a maioria dos valores são nulos, então para economizar memória, vamos transformar essa matriz em uma matriz esparsa.

O método csr_matriz armazenará os dados apenas das posições que possuem valores.

In [78]:
from scipy.sparse import csr_matrix
books_sparse = csr_matrix(book_pivot)

Agora nossa matriz tem um tamanho reduzido e ideal para um treinamento efetivo do modelo NearestNeighbors, que se baseia na distância entre os pontos.

In [79]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors(algorithm='brute')
model.fit(books_sparse)

NearestNeighbors(algorithm='brute')

Vamos agora encontrar os vizinhos mais próximos de um determinado livro. Com model.kneighbors, passamos o índice do livro que queremos encontrar os vizinhos mais próximos em formato de um array 2D, onde o primeiro elemento é o índice do livro e o segundo elemento é o índice do usuário. Além disso passamos o número de vizinhos que queremos encontrar, que no caso é 6.

In [80]:
distance, suggestion = model.kneighbors(book_pivot.iloc[237,:].values.reshape(1,-1),n_neighbors=6)

Vamos agora criar um loop que apresenta todas as sugestões de livros para um determinado usuário.

In [81]:
for i in range(len(suggestion)):
    print(book_pivot.index[suggestion[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')


Nomes dos livros presentes no dataframe

In [82]:
books_name = book_pivot.index

Vamos salvar o modelo, nome dos livros, o data frame final e a matriz pivot em arquivos para que possamos usar em outros notebooks.

In [83]:
import pickle
pickle.dump(model,open('artifacts/model.pkl','wb'))
pickle.dump(books_name,open('artifacts/books_name.pkl','wb'))
pickle.dump(final_rating,open('artifacts/final_rating.pkl','wb'))
pickle.dump(book_pivot,open('artifacts/book_pivot.pkl','wb'))

Agora vamos criar uma função geral que obtem a recomendação de livros para qualquer usuário, como nós.

In [84]:
def recommend_book(book_name):
    book_id = np.where(book_pivot.index==book_name)[0][0]
    distance, suggestion = model.kneighbors(book_pivot.iloc[book_id,:].values.reshape(1,-1),n_neighbors=6)
    
    recommendations = []

    for i in range(len(suggestion)):
        books = book_pivot.index[suggestion[i]]
        for j in books:
            url = final_rating.loc[final_rating['title'].str.contains(j, case=False, regex=False), 'image_url'].values[0]
            recommendation = {'Book': j, 'URL Image': url}
            recommendations.append(recommendation)

    recommendations_df = pd.DataFrame(recommendations)
    return recommendations_df

In [85]:
book_name = 'Harry Potter and the Chamber of Secrets (Book 2)'
recommend_book(book_name)

Book  \
0   Harry Potter and the Chamber of Secrets (Book 2)   
1  Harry Potter and the Prisoner of Azkaban (Book 3)   
2       Harry Potter and the Goblet of Fire (Book 4)   
3     Harry Potter and the Sorcerer's Stone (Book 1)   
4                                          Exclusive   
5                               The Cradle Will Fall   

                                           URL Image  
0  http://images.amazon.com/images/P/0439064872.0...  
1  http://images.amazon.com/images/P/0439136369.0...  
2  http://images.amazon.com/images/P/0439139597.0...  
3  http://images.amazon.com/images/P/043936213X.0...  
4  http://images.amazon.com/images/P/0446604232.0...  
5  http://images.amazon.com/images/P/0440115450.0...